### PR TITLE
chore: refactor JavacOptionsUtils into mostly a builder format

### PR DESCRIPTION
**Pull request by danielmoy-google**
Allows us to get rid of some slow list behavior, helps usability on the
interface so we don't have to remember which methods modify-in-place vs.
remove vs. return new copies vs. some combination of the above.

### CODE REVIEW

File: kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java [hunk 1]
```diff
@@ -61,177 +67,209 @@
...
+    public ModifiableOptions removeUnsupportedOptions() {
+      List<String> previous = internal;
+      internal = new ArrayList<>();
+      ImmutableList<OptionChecker> optionCheckers =
+          ImmutableList.of(JavacTool.create(), new JavacFileManager(new Context(), false, UTF_8));
+      PeekingIterator<String> it =
+          Iterators.peekingIterator(
+              Iterators.filter(
+                  previous.iterator(),
+                  Predicates.not(o -> o.startsWith("-Xlint") || o.startsWith("-Werror"))));
```
Comment (shahms) [inline kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java]: Just do this as an additional option checker?

File: kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java [hunk 2]
```diff
@@ -61,177 +67,209 @@
...
+      internal = new ArrayList<>(original.size());
+      for (int i = 0; i < original.size(); i++) {
+        String opt = original.get(i);
+        boolean matched = false;
+        for (Option o : opts) {
+          if (o.matches(opt)) {
+            matched = true;
+            if (o.hasArg()) {
+              // Skip the argument too.
+              i++;
```
Comment (shahms) [inline kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java]: This is nearly identical (semantically, if not syntactically) with `removeUnsupported` and should probably be unified.

File: kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java [hunk 3]
```diff
@@ -61,177 +67,209 @@
...
+     * Identify the paths associated with the specified option, remove them from the builder, and
+     * return them.
+     */
+    private ImmutableList<String> removeArgumentPaths(Option option) {
+      List<String> original = internal;
+      internal = new ArrayList<>(original.size());
+      ImmutableList.Builder<String> paths = ImmutableList.builder();
+      for (int i = 0; i < original.size(); i++) {
+        if (option.matches(original.get(i))) {
+          if (i + 1 >= original.size()) {
```
Comment (shahms) [inline kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java]: Ditto here on this loop.

File: kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java (hunk 2 above)
Comment (shahms) [reply kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java]: My favorite pattern for dealing with this comes from the JavaFileManager class and is essentially the interface:

interface OptionHandler {
Iterator<String> handleOption(String value, Iterator<String> remaining);
}

// remove
(opt, rest) -> rest
// w/ value
(opt, rest) -> { rest.next(); return rest; }
// keep
(opt, rest) -> Iterator.concat(ImmutableList.of(opt), rest);

etc.

File: kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java (hunk 3 above)
Comment (danielmoy-google) [reply kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java]: sounds good

File: kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java (hunk 2 above)
Comment (danielmoy-google) [reply kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java]: ehh, maybe just the one loop can do it without arbitrary handler.

But, similar logic for different handling of what to do on match / not-match / positionally-matched makes sense still imo.

**Pull request merged**